File: russian-doll-envelopes/russian-doll-envelopes.cpp
class Solution {
    
public:
    int maxEnvelopes(vector<vector<int>>& mat) {
        int n=mat.size(),ans=0;
        sort(mat.begin(),mat.end());
        vector<int>dp(n,1);
        for(int i=0;i<n;i++){
            for(int j=0;j<i;j++){
                if(mat[i][1]>mat[j][1] && mat[i][0]>mat[j][0]){
                    dp[i]=max(dp[i],dp[j]+1);
                }
            }
            ans=max(dp[i],ans);
        }
        return ans;
    }
};


// for(int j=0;j<i;++j){
//                 if(arr[i][0] > arr[j][0] && arr[i][1] > arr[j][1]) dp[i] = max(dp[i], 1+dp[j]);
//             }

File: 354-russian-doll-envelopes/russian-doll-envelopes.py
class Solution:
    def maxEnvelopes(self, e: List[List[int]]) -> int:
        e.sort(key=lambda x: (x[0], -x[1]))
        dp = []
        for num1, num2 in e:
            i = bisect_left(dp,num2)
            if i == len(dp):
                dp.append(num2)
            else:
                dp[i] = num2
        return len(dp)